Add Post Content attributes as a block editor setting

tellthemachines opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/58534

This PR adds the changes from https://github.com/WordPress/gutenberg/pull/45299.

To test, checkout the PR locally and update the "@wordpress/editor" package to its latest version - "13.12.0". Then run the build, and use the Redux dev tools to inspect settings in the state of the `core/block-editor` store.

Edit: I somehow left out that this PR also adds the changes from https://github.com/WordPress/gutenberg/pull/49049 🤦 

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

tellthemachines committed:
Add Post Content attributes as a block editor setting

audrasjb committed:
Fix coding standards issues

audrasjb commented:
Thanks for the PR!
I added one commit to fix a few coding standards issues :)

spacedmonkey commented:
Can you please add some unit tests to this PR.

jeremyfelt commented on src/wp-includes/block-editor.php:
```diff
@@ -361,6 +361,89 @@ function _wp_get_iframed_editor_assets() {
+ * @param array  $blocks     Array of blocks.
+ * @return array Found block, or empty array if none found.
+ */
+function _wp_find_first_block( $block_name, $blocks ) {
```
I think this function could be useful for the general developer community as well. I know I've reinvented this wheel a few times already for extracting images/excerpts from post content on various projects. Can the pseudo-private `_` prefix be removed?

Also, `wp_find_` is used as a prefix only 3 times in core right now. As a developer, I generally expect WordPress functions that return data to use `wp_get_`.

jeremyfelt commented on src/wp-includes/block-editor.php:
```diff
@@ -361,6 +361,89 @@ function _wp_get_iframed_editor_assets() {
+
+	$current_template = get_block_templates( array( 'slug__in' => array( $template_slug ) ) );
+
+	if ( ! empty( $current_template ) ) {
```
I _think_ a basic string check would be quick here and prevent any unnecessary looping through blocks.

```suggestion
	if ( ! empty( $current_template ) && has_block( 'core/post-content', $current_template[0]->content ) ) {
```

That said, I'm not sure how often the template will be very large or not have `core/post-content` in it. :)

tellthemachines replied:
Hmm good point. I think very large templates would be the exception, but it's even less likely that a template used for a post or page _won't_ have `core/post-content` in it. At this point `$current_template` is the post or page template, so perhaps the optimisation isn't necessary? I don't feel strongly about it either way though!

jeremyfelt replied:
That makes sense, I'll take readability over over-optimization. Thanks! :)

tellthemachines committed:
Remove dummy setting

tellthemachines committed:
Rename finding function and make public

tellthemachines committed:
Add test for wp_get_first_block

tellthemachines committed:
Fix linting error

tellthemachines committed:
fix further lint errors

tellthemachines commented:
Thanks for fixing the coding standards stuff @audrasjb! I'm having issues running the composer scripts locally 😩 

I've added a test for the now-renamed `wp_get_first_block`, now working on tests for `_wp_get_post_content_block_attributes`.

tellthemachines committed:
Add test for _wp_get_post_content_block_attributes

tellthemachines commented:
I just added a base test case for `_wp_get_post_content_block_attributes` to check it returns an empty array by default (if there's no post ID and no block template). Not sure how to test this further, as it requires both a post and its block template to exist. I tried setting `$post_ID =1` and `switch_theme( 'block-theme' );` but that doesn't seem to work. I'm probably missing something obvious 😅 any help appreciated!

tellthemachines committed:
Add a test to check for post content attributes

tellthemachines committed:
remove extra tab

tellthemachines commented:
Update: I got it working! Had to add a single template to the test block theme.

mukeshpanchal27 commented on src/wp-includes/block-editor.php:
```diff
@@ -361,6 +361,89 @@ function _wp_get_iframed_editor_assets() {
+function _wp_get_post_content_block_attributes() {
+	$is_block_theme = wp_is_block_theme();
+
+	global $post_ID;
```
Missing docblock for global.

`* @global int $post_ID`

felixarntz replied:
Nit-pick: It would be good to follow the pattern of putting the `global` declaration to the very top of the function code.

mukeshpanchal27 commented on src/wp-includes/block-editor.php:
```diff
@@ -529,6 +612,10 @@ function get_block_editor_settings( array $custom_settings, $block_editor_contex
 		),
 	);
 
+	if ( ! empty( _wp_get_post_content_block_attributes() ) ) {
+		$editor_settings['postContentAttributes'] = _wp_get_post_content_block_attributes();
```
`_wp_get_post_content_block_attributes()` call multiple time if condition true instead define a variable and use it.

tellthemachines replied:
Makes sense!

mukeshpanchal27 reviewed (commented):
Thanks @tellthemachines, Left some nitpick feedback.

felixarntz commented on src/wp-includes/block-editor.php:
```diff
@@ -361,6 +361,89 @@ function _wp_get_iframed_editor_assets() {
+ * @param array  $blocks     Array of blocks.
+ * @return array Found block, or empty array if none found.
+ */
+function wp_get_first_block( $block_name, $blocks ) {
```
This may be just a personal preference, but I would recommend to reverse the order of those parameters. Here's why:
* If `$blocks` had some default value and was optional, it would make sense that it's the 2nd parameter. But given that it has to be provided, I think it's more intuitive to put it first because that's the source data where to look in.
* It's intuitive that for such a function you surely have to pass the block name / identifier, but whether or not you have to pass the source to look in is less obvious (e.g. in a function like `get_post( $post_id )` you don't have to provide a source), which to me is another reason to clarify that by putting that source parameter first.
* Last but not least, it's also a bit closer in the order to common code patterns like `someArray.find( value )`, basically "blocks -> find -> block_name".

tellthemachines replied:
Sounds good!

felixarntz commented on src/wp-includes/block-editor.php:
```diff
@@ -361,6 +361,89 @@ function _wp_get_iframed_editor_assets() {
+/**
+ * Adds styles and __experimentalFeatures to the block editor settings.
+ *
+ * @since 6.3.0
```
```suggestion
 * @since 6.3.0
 * @access private
```

felixarntz commented on src/wp-includes/block-editor.php:
```diff
@@ -361,6 +361,89 @@ function _wp_get_iframed_editor_assets() {
+ *
+ * @return array Post Content block attributes or empty array if they don't exist.
+ */
+function _wp_get_post_content_block_attributes() {
```
While WP core uses some underscore-prefixed functions, several seasoned contributors have advised in the past to not do that anymore. I would therefore suggest to replace the `_wp` here with just `wp` and instead rely just on an `@access private` annotation which is the more appropriate way to clarify its private nature.

tellthemachines replied:
Ooh ok I was't aware of that.

felixarntz commented on src/wp-includes/block-editor.php:
```diff
@@ -361,6 +361,89 @@ function _wp_get_iframed_editor_assets() {
+}
+
+/**
+ * Adds styles and __experimentalFeatures to the block editor settings.
```
Is this description correct? It looks to me like that's not what the function is doing.

tellthemachines replied:
Good point, that was accurate for the Gutenberg function which is a block editor settings filter, but not for this one.

felixarntz reviewed (approved):
@tellthemachines Looks solid to me, +1 to having `wp_get_first_block()` as a generally available function.

I left a few small comments below, but none strictly a blocker (except probably the two documentation bits).

tellthemachines committed:
Change order of wp_get_first_block parameters

tellthemachines committed:
Update function name and meta

tellthemachines committed:
Avoid calling function twice

tellthemachines committed:
Add test to check for post content attributes in editor settings

tellthemachines committed:
Remove extra whitespace

tellthemachines commented:
Thanks for the feedback folks!

Question on the tests: I added an extra assertion to `test_get_block_editor_settings_theme_json_settings` to check for the new setting, and in doing so realised that the test is now relying on the global set above in `test_wp_get_post_content_block_attributes` to pass. This isn't great as it could also alter results of any tests running after. Do we have any best practices for handling globals in tests? Should they be unset after each test?

tellthemachines committed:
Test: handle globals in setup/teardown

tellthemachines committed:
Fix lint errors

tellthemachines commented:
> Do we have any best practices for handling globals in tests? Should they be unset after each test?

Update: @hellofromtonya has pointed me in the right direction and I've updated the tests accordingly.

tellthemachines committed:
try again

tellthemachines committed:
fix spacing

andrewserong commented on tests/phpunit/tests/blocks/editor.php:
```diff
@@ -27,12 +27,18 @@ public function set_up() {
 		do_action( 'rest_api_init', $wp_rest_server );
+
+		global $post_ID;
+		$this->orig_post_id = $post_ID;
```
Looks like this is throwing errors in the PHP 8.2 actions:

<img width="641" alt="image" src="https://github.com/WordPress/wordpress-develop/assets/14988353/c66c647a-0d35-414f-a8fb-53de0ae582e9">

Would the fix be to define the `orig_post_id` property at the top of this class?

tellthemachines replied:
I wasn't sure, so I copied what's being done for the `$wp_rest_server` global and set it to null on teardown. It shouldn't matter because `$post_ID` shouldn't be set to any value at this point (if it was, `test_wp_get_post_content_block_attributes` wouldn't fail without it being set explicitly).

andrewserong replied:
Looks good to me!

tellthemachines committed:
Fix deprecation errors

felixarntz commented on src/wp-includes/block-editor.php:
```diff
@@ -361,6 +361,92 @@ function _wp_get_iframed_editor_assets() {
+ * @return array Post Content block attributes or empty array if they don't exist.
+ */
+function wp_get_post_content_block_attributes() {
+	$is_block_theme = wp_is_block_theme();
+
+	global $post_ID;
```
See https://github.com/WordPress/wordpress-develop/pull/4614/files#r1232517228, could you move this to the top please?
```suggestion
	global $post_ID;

	$is_block_theme = wp_is_block_theme();
```

tellthemachines replied:
Oh whoops, missed that! Fixed.

felixarntz reviewed (approved):
Thanks for the updates @tellthemachines, LGTM! Just the one nit-pick remaining, but not a big deal. Feel free to update before commit.

tellthemachines committed:
Move global to top of function.

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

mukeshpanchal27 reviewed (approved):
Thanks @tellthemachines the update. Look good to me.

mukeshpanchal27 commented:
@tellthemachines Re-run the failed job it show `Navigation timeout of 30000 ms exceeded` .

tellthemachines commented:
Thanks again for the reviews, everyone!

Committed in [r55955](https://core.trac.wordpress.org/changeset/55955) / [b35b1fc](https://github.com/WordPress/wordpress-develop/commit/b35b1fcc869dc9c09ec184801ec9b498d16d9af6).
